### PR TITLE
Allow await when it is not in AsyncArrowHead

### LINKED ISSUE #10988: await should be allowed when it is not in AsyncArrowHead
## Bug Report

- [ ] I would like to work on a fix!

**Current Behavior**
A clear and concise description of the behavior.
The following snippet will throw SyntaxError: Await cannot be used as name inside an async function (1:1)

**Input Code**
- REPL or Repo link if applicable: [REPL](https://babeljs.io/repl#?browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=BQQw7iCWAuCUAEBeAfPA3gXyA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.8.4&externalPlugins=)

```js
(await) => {}
```

**Expected behavior/code**
It should not throw because it is not in async arrow parameters, which means `await` is a plain identifier and should be allowed in ArrowParameters because it does not have `[Await]` production parameter.
https://tc39.es/ecma262/#sec-arrow-function-definitions

**Babel Configuration**
REPL, no plugins are enabled

**Environment**

REPL

**Possible Solution**

Should revisit how `awaitPos` is tracked in the parser. You can pay attention to
https://github.com/babel/babel/blob/a0a9c64a470acec77bec1da0b4bceec4ccf7f446/packages/babel-parser/src/parser/expression.js#L2172-L2175

where `awaitPos` is updated when it may be in arrow parameters. It was used to track `awaitPos` on cases like
```js
async ( x = (await) => {}) => {}
```
however it becomes overkill when it is _not_ in _async_ arrow parameters. Consider add a new parser state `maybeInAsyncArrowHead` when parsing ambiguous async arrow

https://github.com/babel/babel/blob/master/packages/babel-parser/src/parser/expression.js#L649

It should be similar to `maybeAsyncArrow` but it should have longer lifecycle -- `maybeInAsyncArrowHead` should be true during the whole async arrow parameters. i.e.
```js
// top level
async ( // maybeInAsyncArrowHead : true
  x = ( // maybeInAsyncArrowHead : true
    await // register the `awaitPos`
  ) => async ( // oldMaybeInAsyncArrowHead should be saved, new maybeInAsyncArrowHead should be true
    y
  ) => {}
); // we does not see an arrow, so the parsed ambiguous async arrow is indeed a CallExpression, reset `awaitPos`.
```

When we are sure we are not parsing an async arrow, i.e. `async(foo)`, we should reset `awaitPos` if `state.maybeInAsyncArrowHead` is `false`.

Then we can replace `maybeInArrowParameters` by `maybeInAsyncArrowHead` it should not record `awaitPos` for `(await) => {}`.

### PR DESCRIPTION
**Pull request by arku**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #10988 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR fixes the parser error thrown for the input `(await) => {}`. I tried to fix the other cases mentioned in the issue description and after struggling with delaying the await position check in case of async call expressions, I reached out to @JLHwung for help. He added that the delay check is indeed difficult and it is best to refactor the await position tracking with a scope based recording mechanism that is non-trivial.

### CODE REVIEW

File: packages/babel-parser/src/tokenizer/state.js [hunk 1]
```diff
@@ -59,6 +59,7 @@ export default class State {
   // Flags to track
   inParameters: boolean = false;
   maybeInArrowParameters: boolean = false;
+  maybeInAsyncArrowHead: boolean = false;
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/tokenizer/state.js]: It's not clear to me what the difference between these two flags should be. Isn't the arrow head its parameters?

Comment (JLHwung) [reply packages/babel-parser/src/tokenizer/state.js]: The `maybeInAsyncArrowHead` lives across function declarations.
```js
async (foo = function (await) {}) => {}
```
When parsing `await`, `maybeInAsyncArrowHead` is `true` but `maybeInArrowParameters` is `false`.

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/tokenizer/state.js]: We should add a comment in the code explaining it 👍

Comment (arku) [reply packages/babel-parser/src/tokenizer/state.js]: @nicolo-ribaudo Done. Let me know if we can make it better.

File: packages/babel-parser/src/parser/expression.js [hunk 2]
```diff
@@ -560,10 +560,15 @@ export default class ExpressionParser extends LValParser {
       stop: false,
     };
     do {
+      const oldMaybeInAsyncArrowHead = this.state.maybeInAsyncArrowHead;
+      if (state.maybeAsyncArrow) {
+        this.state.maybeInAsyncArrowHead = state.maybeAsyncArrow;
+      }
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/expression.js]: Nit: inside the `if` `state.maybeAsyncArrow` is always `true`.

```suggestion
      if (state.maybeAsyncArrow) this.state.maybeInAsyncArrowHead = true;
```

Comment (arku) [reply packages/babel-parser/src/parser/expression.js]: @nicolo-ribaudo Ah, yeah. Just updated the PR.

**Pull request merged**